### PR TITLE
CRUD-functionality implemented

### PR DESCRIPTION
**Pull request by sirtoobii**
As the title says, this version implements all basic CRUD-functionality

What is still missing:
- [x] Send email with client configuration to users
- [x] Display client config (e.g using a QR-code)
- [x] Input validation - especially for the IP-Address field
- [x] Search functionality
- [x] Configuration backups(!)
- [x] Docs
- [x] Integration in a "real" setup -> How to allow modification of wireguard configurations without compromising the overall system security?
- [x] Fix automated tests(?)
- [x] Pretty error messages

### CODE REVIEW

File: bin/wgwrangler.pl [hunk 1]
```diff
@@ -1,7 +1,8 @@
 #!/usr/bin/env perl
 
 use lib qw(); # PERL5LIB
-use FindBin;use lib "$FindBin::RealBin/../lib";use lib "$FindBin::RealBin/../thirdparty/lib/perl5"; # LIBDIR
```
Comment (oetiker) [inline bin/wgwrangler.pl]: the point of having this on a single line is that the # LIBDIR comment at the end of the line causes the entire line to be replaced with an absolute path upon installation

File: conftools/install-sh [hunk 2]
```diff
@@ -1,7 +1,7 @@
 #!/bin/sh
```
Comment (oetiker) [inline conftools/install-sh]: the whole conftools folder should not be in git ... as it is generated ...

File: lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+sub getAllFieldValues {
+    my $self = shift;
+    my $args = shift;
+    my $formData = shift;
+
+    my $data = $self->app->wireguardModel->gen_key_pair();
+
+    if ($formData->{currentFormData}{interface}) {
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm]: is it safe to accept just any input from the outside at this point ? a bad actor could be feeding random data to the api

File: lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+};
+
+has actionCfg => sub {
+    my $self = shift;
+
+    my $handler = sub {
+        my $self = shift;
+        my $args = shift;
+
+        my $interface = $args->{interface};
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm:296]: is it save to accept just any incoming data ?

File: lib/WGwrangler/GuiPlugin/WireguardShow.pm [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+    my $self = shift;
+    my $args = shift;
+    my $filter = $self->_getFilter($args->{formData}{wg_interface});
+    return $self->app->wireguardModel->get_peer_count($filter);
+}
+
+sub getTableData {
+    my $self = shift;
+    my $args = shift;
+    my $filter = $self->_getFilter($args->{formData}{song_title});
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardShow.pm]: leftover naming ?

File: lib/WGwrangler/IPmanager.pm [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    else {
+        return ();
+    }
+}
+
+
+sub _generate_ipv4_list($network_id, $subnet_size, $ref_ip_database) {
+    # thanks to https://www.perl.com/article/creating-ip-address-tools-from-scratch/
+
+    my $n_ips = 0;
```
Comment (oetiker) [inline lib/WGwrangler/IPmanager.pm]: there are some pretty powerful perl modules for managing ip addresses around ... including ipv6 support

File: lib/WGwrangler/Model/WireguardDataAdapter.pm [hunk 7]
```diff
@@ -0,0 +1,188 @@
+package WGwrangler::Model::WireguardDataAdapter;
+use strict;
+use warnings FATAL => 'all';
+use experimental 'signatures';
+use Mojo::JSON qw(true false);
+
+use base 'Exporter';
+our @EXPORT = qw(get_peer_table_data get_peer_count update_peer_data sort_table_data commit_changes get_section_data disable_peer get_interface_selection);
```
Comment (oetiker) [inline lib/WGwrangler/Model/WireguardDataAdapter.pm]: I would suggest to only export upon request using `EXPORT_OK` to not randomply pollute the users namespace

File: lib/WGwrangler/User.pm [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package WGwrangler::User;
+use strict;
+use warnings FATAL => 'all';
+
+#use Mojo::Base 'Mojolicious';
+
+our @ISA = qw (CallBackery::User);
```
Comment (oetiker) [inline lib/WGwrangler/User.pm]: In mojo land, just do:

```suggestion
use Mojo::Base 'CallBackery::User';
```

**Review by oetiker — CHANGES_REQUESTED** [review]

File: lib/WGwrangler/IPmanager.pm (hunk 6 above)
Comment (sirtoobii) [reply lib/WGwrangler/IPmanager.pm]: Indeed, migrated to Net::IP - however, I had to drop ip suggestions for now...

File: bin/wgwrangler-source-mode.sh [hunk 9]
```diff
@@ -1,4 +1,6 @@
 #!/bin/sh
 export MOJO_MODE=development
 export MOJO_LOG_LEVEL=debug
-exec `dirname $0`/wgwrangler.pl prefork --listen 'http://*:7192'
+export IS_TESTING=1
```
Comment (oetiker) [inline bin/wgwrangler-source-mode.sh]: I would suggest to use namespacing ... 

`WGWRANGLER_IS_TESTING`

File: etc/wgwrangler.yaml [hunk 10]
```diff
@@ -1,15 +1,23 @@
-
 BACKEND:
```
Comment (oetiker) [inline etc/wgwrangler.yaml]: I would not distribute the 'non dist' version of the config file

File: lib/WGwrangler/GuiPlugin/CommitMessageForm.pm [hunk 11]
```diff
@@ -0,0 +1,128 @@
+package WGwrangler::GuiPlugin::CommitMessageForm;
+use Mojo::Base 'CallBackery::GuiPlugin::AbstractForm';
+use CallBackery::Translate qw(trm);
+use CallBackery::Exception qw(mkerror);
+use Mojo::JSON qw(true false);
+use experimental 'signatures';
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/CommitMessageForm.pm]:
```suggestion
use Mojo::Base 'CallBackery::GuiPlugin::AbstractForm', -signatures;
use CallBackery::Translate qw(trm);
use CallBackery::Exception qw(mkerror);
use Mojo::JSON qw(true false);
```

File: lib/WGwrangler/GuiPlugin/CommitMessageForm.pm [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+=head2 formCfg
+
+Returns a Configuration Structure for the Song Entry Form.
+
+=cut
+
+
+has screenOpts => sub {
+    my $self = shift;
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/CommitMessageForm.pm]:
```suggestion
has screenOpts => sub ($self) {
```

File: lib/WGwrangler/GuiPlugin/CommitMessageForm.pm [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+                height    => 120
+            },
+            addProps => {
+                edge => 'west',
+            }
+        }
+    }
+};
+
+has formCfg => sub($self) {
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/CommitMessageForm.pm]:
```suggestion
has formCfg => sub ($self) {
```

File: lib/WGwrangler/GuiPlugin/CommitMessageForm.pm [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+            set    => {
+                placeholder => trm('Briefly describe your changes'),
+                required    => true
+            }
+        },
+    ];
+};
+
+has actionCfg => sub {
+    my $self = shift;
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/CommitMessageForm.pm]:
```suggestion
has actionCfg => sub ($self) {
```

File: lib/WGwrangler/GuiPlugin/CommitMessageForm.pm [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+};
+
+sub getAllFieldValues {
+    my $self = shift;
+    my $args = shift;
+    return [];
+}
+
+has checkAccess => sub {
+    my $self = shift;
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/CommitMessageForm.pm]:
```suggestion
has checkAccess => sub ($self) {
```

File: lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm [hunk 16]
```diff
@@ -0,0 +1,461 @@
+package WGwrangler::GuiPlugin::WireguardAddPeerForm;
+use Mojo::Base 'CallBackery::GuiPlugin::AbstractForm';
+use WGwrangler::Model::MailHandler;
+use CallBackery::Translate qw(trm);
+use CallBackery::Exception qw(mkerror);
+use Mojo::JSON qw(true false);
+use experimental 'signatures';
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm]:
```suggestion
use Mojo::Base 'CallBackery::GuiPlugin::AbstractForm', -signatures;
use WGwrangler::Model::MailHandler;
use CallBackery::Translate qw(trm);
use CallBackery::Exception qw(mkerror);
use Mojo::JSON qw(true false);
```

File: lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm [hunk 17]
```diff
@@ -0,0 +1,461 @@
...
+
+WGwrangler::GuiPlugin::WireguardPeerForm - Peer Edit form
+
+=head1 DESCRIPTION
+
+Peer edit form
+
+=cut
+
+has 'mailHandler' => sub($self) {
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm]:
```suggestion
has 'mailHandler' => sub ($self) {
```

File: lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm [hunk 18]
```diff
@@ -0,0 +1,461 @@
...
+=cut
+
+=head2 formCfg
+
+Returns a Configuration Structure for the Wireguard Show form
+
+=cut
+
+
+has formCfg => sub($self) {
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm]:
```suggestion
has formCfg => sub ($self) {
```

File: lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm [hunk 19]
```diff
@@ -0,0 +1,461 @@
...
+    return [
+        # interface selection
+        {
+            key              => 'interface',
+            label            => trm('Interface'),
+            widget           => 'selectBox',
+            triggerFormReset => true,
+            # interestingly the required property is not enough here...
+            validator        => sub($value, $parameter, $formData) {
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm]:
```suggestion
            validator        => sub ($value, $parameter, $formData) {
```

File: lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm [hunk 20]
```diff
@@ -0,0 +1,461 @@
...
+            widget           => 'checkBox',
+            triggerFormReset => true,
+        },
+        # Header
+        {
+            key    => 'header_client',
+            widget => 'header',
+            label  => trm('Client specific configuration'),
+            set    => {
+                visibility => $self->{args}{currentFormData}{show_advanced} ? 'visible' : 'excluded',
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm]: I would save this in a variable to reuse

File: lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm [hunk 21]
```diff
@@ -0,0 +1,461 @@
...
+    }
+    else {
+        $data->{'config_preview'} = 'Select an interface first';
+    }
+
+    return $data;
+}
+
+has checkAccess => sub {
+    my $self = shift;
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardAddPeerForm.pm]:
```suggestion
has checkAccess => sub ($self)  {
```

File: lib/WGwrangler/GuiPlugin/WireguardEditPeerForm.pm [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+            set    => {
+                readOnly => true,
+                required => true
+            }
+        },
+        {
+            key    => 'device',
+            label  => trm('Device'),
+            widget => 'text',
+            validator        => sub($value, $parameter, $formData) {
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardEditPeerForm.pm]:
```suggestion
            validator        => sub ($value, $parameter, $formData) {
```

File: lib/WGwrangler/GuiPlugin/WireguardEditPeerForm.pm [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+            set    => {
+                readOnly => true,
+                required => true,
+            }
+        },
+        {
+            key       => 'name',
+            label     => trm('Name'),
+            widget    => 'text',
+            validator => sub($value, $parameter, $formData) {
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardEditPeerForm.pm]:
```suggestion
            validator => sub ($value, $parameter, $formData) {
```

File: lib/WGwrangler/GuiPlugin/WireguardEditPeerForm.pm [hunk 24]
```diff
@@ -0,0 +1,193 @@
...
+            },
+            set       => {
+                required => true,
+            },
+        },
+        {
+            key       => 'allowed-ips',
+            label     => trm('Allowed-IPs'),
+            widget    => 'text',
+            validator => sub($value, $parameter, $formData) {
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardEditPeerForm.pm]:
```suggestion
            validator => sub ($value, $parameter, $formData) {
```

File: lib/WGwrangler/GuiPlugin/WireguardEditPeerForm.pm [hunk 25]
```diff
@@ -0,0 +1,193 @@
...
+            set    => {
+                placeholder => trm('Some extra infos about this peer'),
+            }
+        },
+    ];
+};
+
+has actionCfg => sub {
+
+    my $handler = sub($self, $args) {
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardEditPeerForm.pm]:
```suggestion
    my $handler = sub ($self, $args) {
```

File: lib/WGwrangler/GuiPlugin/WireguardEditPeerForm.pm [hunk 26]
```diff
@@ -0,0 +1,193 @@
...
+};
+
+sub getAllFieldValues {
+    my $self = shift;
+    my $args = shift;
+    return $self->app->wireguardModel->get_section_data($args->{selection}{interface}, $args->{selection}{'public-key'});
+}
+
+has checkAccess => sub {
+    my $self = shift;
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardEditPeerForm.pm]:
```suggestion
has checkAccess => sub ($self) {
```

File: lib/WGwrangler/GuiPlugin/WireguardShow.pm [hunk 27]
```diff
@@ -0,0 +1,390 @@
+package WGwrangler::GuiPlugin::WireguardShow;
+use Mojo::Base 'CallBackery::GuiPlugin::AbstractTable';
+use CallBackery::Translate qw(trm);
+use CallBackery::Exception qw(mkerror);
+use Mojo::JSON qw(true false);
+use Wireguard::WGmeta::Utils;
+use experimental 'signatures';
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardShow.pm]:
```suggestion
use Mojo::Base 'CallBackery::GuiPlugin::AbstractTable', -signatures;
use CallBackery::Translate qw(trm);
use CallBackery::Exception qw(mkerror);
use Mojo::JSON qw(true false);
use Wireguard::WGmeta::Utils;
```

File: lib/WGwrangler/GuiPlugin/WireguardShow.pm [hunk 28]
```diff
@@ -0,0 +1,390 @@
...
+
+WGwrangler::GuiPlugin::WireguardShow - Shows contents of WG configs
+
+=head1 METHODS
+
+All the methods of L<CallBackery::GuiPlugin::AbstractTable> plus:
+
+=cut
+
+has formCfg => sub($self) {
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardShow.pm]:
```suggestion
has formCfg => sub ($self) {
```

File: lib/WGwrangler/GuiPlugin/WireguardShow.pm [hunk 29]
```diff
@@ -0,0 +1,390 @@
...
+            set    => {
+                placeholder => 'name, interface, email, ip, public-key, device',
+                enabled     => true
+            },
+        },
+    ]
+};
+
+has tableCfg => sub {
+    my $self = shift;
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardShow.pm]:
```suggestion
has tableCfg => sub ($self) {
```

File: lib/WGwrangler/GuiPlugin/WireguardVersions.pm [hunk 30]
```diff
@@ -0,0 +1,179 @@
+package WGwrangler::GuiPlugin::WireguardVersions;
+use Mojo::Base 'CallBackery::GuiPlugin::AbstractTable';
+use CallBackery::Translate qw(trm);
+use CallBackery::Exception qw(mkerror);
+use Mojo::JSON qw(true false);
+use Wireguard::WGmeta::Utils;
+use experimental 'signatures';
```
Comment (oetiker) [inline lib/WGwrangler/GuiPlugin/WireguardVersions.pm]:
```suggestion
use Mojo::Base 'CallBackery::GuiPlugin::AbstractTable', -signatures;
use CallBackery::Translate qw(trm);
use CallBackery::Exception qw(mkerror);
use Mojo::JSON qw(true false);
use Wireguard::WGmeta::Utils;
```

File: lib/WGwrangler/Model/IPmanager.pm [hunk 31]
```diff
@@ -0,0 +1,268 @@
...
+
+sub new($class) {
+    my $self = {
+        'interface_ranges' => {},
+        'ip_meta_info'     => {},
+        'acquired_ips'     => {},
+    };
+    bless $self, $class;
+    return $self;
+}
```
Comment (oetiker) [inline lib/WGwrangler/Model/IPmanager.pm]: ich wuerde empfehlen das model mit Mojo::base zu bauen ... gleich wie das Email module ...

File: lib/WGwrangler/User.pm [hunk 32]
```diff
@@ -0,0 +1,20 @@
+package WGwrangler::User;
```
Comment (oetiker) [inline lib/WGwrangler/User.pm]: hmmm :-)

**Review by oetiker — CHANGES_REQUESTED** [review]

**Pull request merged**